platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/DefaultJavaToolchainProvisioningService.java

```diff
@@ -143,16 +143,29 @@ private File provisionInstallation(JavaToolchainSpec spec, URI uri, Collection<A
             try {
                 File downloadFolder = cacheDirProvider.getDownloadLocation();
                 ExternalResource resource = wrapInOperation("Examining toolchain URI " + uri, () -> downloader.getResourceFor(uri, authentications));
-                File archiveFile = new File(downloadFolder, getFileName(uri, resource));
+                File archiveFile = new File(downloadFolder, getUniqueFileName(uri, resource, spec));
                 final FileLock fileLock = cacheDirProvider.acquireWriteLock(archiveFile, "Downloading toolchain");
+                boolean canSkipDownload = archiveFile.exists();
                 try {
-                    if (!archiveFile.exists()) {
+                    if (!canSkipDownload) {
                         wrapInOperation("Downloading toolchain from URI " + uri, () -> {
                             downloader.download(uri, archiveFile, resource);
                             return null;
                         });
                     }
-                    return wrapInOperation("Unpacking toolchain archive " + archiveFile.getName(), () -> cacheDirProvider.provisionFromArchive(spec, archiveFile, uri));
+                    try {
+                        return wrapInOperation("Unpacking toolchain archive " + archiveFile.getName(), () -> cacheDirProvider.provisionFromArchive(spec, archiveFile, uri));
+                    } catch (Exception e) {
+                        if (canSkipDownload) { // re-download and retry in case the archive is corrupted
```

Comment:
🤔 Same comments here.  Also, if we're already logging statuses of "Unpacking..." and "Downloading...", should we log the fact we hit an exception, which is why we're downloading again here?